TrafficCapture/dockerSolution/src/main/docker/migrationConsole/lib/console_link/console_link/models/migration.py

```diff
@@ -0,0 +1,110 @@
+from console_link.logic.osi_migration import OSIMigrationLogic, OpenSearchIngestionMigrationProps
+from console_link.models.cluster import Cluster
+from typing import Dict
+from enum import Enum
+from cerberus import Validator
+
+OSI_SCHEMA = {
+    'pipeline_role_arn': {
+        'type': 'string',
+        'required': True
+    },
+    'vpc_subnet_ids': {
+        'type': 'list',
+        'required': True,
+        'schema': {
+            'type': 'string',
+        }
+    },
+    'security_group_ids': {
+        'type': 'list',
+        'required': True,
+        'schema': {
+            'type': 'string',
+        }
+    },
+    'aws_region': {
+        'type': 'string',
+        'required': True
+    },
+    'pipeline_name': {
+        'type': 'string',
+        'required': False
+    },
+    'index_regex_selection': {
+        'type': 'list',
+        'required': False,
+        'schema': {
+            'type': 'string',
+        }
+    },
+    'log_group_name': {
+        'type': 'string',
+        'required': False
+    },
+    'tags': {
+        'type': 'list',
+        'required': False,
+        'schema': {
+            'type': 'string',
+        }
+    }
+}
+
+
+class MigrationType(str, Enum):
+    OSI_HISTORICAL_MIGRATION = "OSI_HISTORICAL_MIGRATION"
+
+
+class Migration():
+    """
+    A base migration manager.
+    """
+
+    def create(self):
+        raise NotImplementedError
+
+    def start(self):
+        raise NotImplementedError
+
+    def stop(self):
+        raise NotImplementedError
+
+    def get_status(self):
+        raise NotImplementedError
+
+
+class OpenSearchIngestionMigration(Migration):
+    """
+    A migration manager for an OpenSearch Ingestion pipeline.
+    """
+
+    def __init__(self, config: Dict, source_cluster: Cluster, target_cluster: Cluster) -> None:
+        v = Validator(OSI_SCHEMA)
+        if not v.validate(config):
+            raise ValueError("Invalid config file for OpenSearchIngestion migration", v.errors)
+        self.osi_props = OpenSearchIngestionMigrationProps(config=config)
+        self.migration_logic = OSIMigrationLogic()
```

Comment:
this is interesting. I'm not opposed to it, but I had been picturing logic imports models, not models imports logic.  I'm game to play with it for now and see what we end up with. The risk of having both directions right now is that we're going to run into circular imports at some point, but we can play with it until we hit that.

Reply:
One point of consideration, I guess -- are there any contexts outside of this class that will use the OSIMigrationLogic? If not, I'd tend to make them internal methods of this class.

My thought was that `logic` would be used to make sure that the users of the library could remain as thin as possible--so `catIndices` goes in the `logic` module because 1. it means that callers don't have to know anything about the exact endpoint or how to call it, but 2. it would be the same logic for all `Cluster`s, regardless of type, so there's no reason to re-implement it multiple times in each class.

I don't know if that's the optimal approach, very open to discussion on it. But I'd tend to say that if the OSIMigrationLogic stuff is more "internal" it should go inside this class or into a `osi_utils` sort of spot.

Reply:
Per discussion, have moved all implementation details out of `logic` and they now reside in `osi_utils`. Still expecting a `middleware` layer that will create this `OpenSearchIngestionMigration` from the raw input it receives